### DISCUSSION TITLE
Next can't be built when using API routes (Next 9.4)

### DISCUSSION THREAD

**Discussion by avxkim** [vercel/next.js#12785] [Help] [answered]
I'm using API routes feature from Next: https://nextjs.org/docs/api-routes/introduction

But faced an issue when doing `next build` because one of my pages are making request to that API route, which obviously isn't available during build, because there was no `next start` command before `next build`, so no server is running. How to deal with this?

Page with request to API route:

```
export const getStaticProps: GetStaticProps = async () => {
  const dicts = await api.request(DICTIONARIES_QUERY);
  return {
    props: {
      dictionaries: {
        total: dicts.dictionaries.total,
        items: dicts.dictionaries.items,
      },
    },
  };
};
```

fetch client:

```
export const api = new GraphQLClient(`http://localhost:3000/api/graphql`);
```

Error during `next build`:

```
Error occurred prerendering page "/dictionary". Read more: https://err.sh/next.js/prerender-error:
FetchError: request to http://localhost:3000/api/graphql failed, reason: connect ECONNREFUSED 127.0.0.1:3000
```

**Comment by brianlovin** [comment, marked as answer]
The solution to this is to basically use the API functions directly in `getStaticProps` - see https://github.com/zeit/next.js/discussions/11210#discussioncomment-1693 for more info

**Reply by avxkim** [reply]
Doesn't make sense to me, honestly, he's calling one route from another, but i call an API route from a PAGE (COMPONENT), not another route.

**Reply by brianlovin** [reply]
With GraphQL this is just harder – your production API endpoint will not know about any changes to your schema that you made locally, so build time queries will fail. This thread might help you, as I encountered the same problem: https://github.com/zeit/next.js/discussions/11285#discussioncomment-8721 - basically I had to use a `SchemaLink` instead of an `HTTPLink` so that queries could execute within the code itself while on the server, and then subsequent clients would use an HTTP request to my endpoint.

**Reply by avxkim** [reply]
As i understand, we only have 2 choices:

1. using `getServerSideProps` instead of `getStaticProps`.
2. Using direct db query inside `getStaticProps` or as you said above, getting data from GraphQL without http link.

Since i'm using `apollo-server-micro`, i just read documentation, and didn't found a way to query GraphQL without http requests.

**Reply by Timer** [reply]
> didn't found a way to query GraphQL without http requests.

Please see these docs for fetching data without HTTP requests:
https://www.apollographql.com/docs/link/links/schema/

**Reply by avxkim** [reply]
@Timer thanks, i'm using [graphql-request](https://github.com/prisma-labs/graphql-request) to fetch data on the client, should i switch to apollo-client in order to make it work with schema link?

**Reply by brianlovin** [reply]
@webcoderkz here is a PR where I managed to implement the SchemaLink/HttpLink switch, based on if it's being run during SSG or not: https://github.com/brianlovin/brian-lovin-next/commit/e553fa9383b55771fb2cecbb228fc521de6ee28b – note that yeah, I had to also move away from `graphql-request` to do this

**Reply by avxkim** [reply]
i just went through apollo documentation, not that bad, imo.

**Reply by benjaminpearson** [reply]
I ran into the same issue and this is how I solved it for the project I'm working on. Basically exporting the GraphQL schema and using it to drive both the API route and to do static generation. https://github.com/vercel/next.js/pull/13202
